:alien: Modernize apt key handling

LeSpocky opened the pull request:
apt-key is deprecated.  Nowadays you put the GPG key of third party apt repo in a local file and reference that with "signed-by" from your sources.list snippet.  For not blindly trusting a key downloaded from some arbitrary URL, the current one is put to this repo now, it is this one:

    pub   rsa4096 2017-02-22 [SCEA]
          9DC858229FC7DD38854AE2D88D81803C0EBFCD88
    uid           Docker Release (CE deb) <docker@docker.com>
    sub   rsa4096 2017-02-22 [S]

This fixes at least the following warning I get by mail from one of the hosts I'm using this role on:

    W: https://download.docker.com/linux/debian/dists/bookworm/InRelease: Schlüssel ist im veralteten
    Schlüsselbund trusted.gpg gespeichert (/etc/apt/trusted.gpg), siehe den Abschnitt MISSBILLIGUNG in
    apt-key(8) für Details.

Link: https://manpages.debian.org/bookworm/apt/apt-key.8.en.html#DEPRECATION
Link: https://michael-prokop.at/blog/2021/02/16/how-to-properly-use-3rd-party-debian-repository-signing-keys-with-apt/
Link: https://wiki.debian.org/DebianRepository/UseThirdParty
Link: https://docs.docker.com/engine/install/debian/#install-using-the-repository

penguineer commented on tasks/main.yml:
```diff
@@ -27,19 +27,32 @@
+      ansible.builtin.copy:
+        src: "files{{ docker_apt_keyrings_dir }}/docker.asc"
+        dest: "{{ docker_apt_keyrings_dir }}"
+        mode: '0644'
```
As APT keys are frequently renewed, in an update run there might be a change of keys without a change of repository listing.

To reflect this change in the APT index, `register: apt_repo`is needed here as well.

LeSpocky replied:
This is addressed in v2, which I just force-pushed.

penguineer reviewed (changes requested):
There is a hidden  change in the PR that needs to be addressed:

Previously the APT key has been downloaded. Now the key is part of the role configuration. As keys are frequently renewed, this changes the maintenance process, as frequent checks and renewals for the key file are needed.

To mitigate this issue I propose one of the following:
1. Change to downloading the key instead of storing the downloaded file in the role.
2. Add a GitHub workflow to frequently check if the key is still up to date and create a PR if it isn't.

If key validation after the download is desired, an (optional!) argument for a key fingerprint can be added.

The main issue is that with this change the role becomes tied to a specific key in the repository, which has a huge influence on the maintenance cycle.

LeSpocky commented:
> There is a hidden  change in the PR that needs to be addressed:
> 
> Previously the APT key has been downloaded. Now the key is part of the role configuration. As keys are frequently renewed, this changes the maintenance process, as frequent checks and renewals for the key file are needed. The repository is now tied to a specific key.

Yeah, this would probably be acceptable for a site project, but not for a role.

> To mitigate this issue I propose one of the following:
> 1. Change to downloading the key instead of storing the downloaded file in the role.

This would be the same behaviour as before.

> 2. Add a GitHub workflow to frequently check if the key is still up to date and create a PR if it isn't.
> 
> If key validation after the download is desired, an (optional!) argument for a key fingerprint can be added.

The last one is actually a good idea. So the role can stay as is, and users who actually care, can set the option.

LeSpocky commented:
v2:
- split into three commits
    1. introduce variable for key download URL (no functional change)
    2. migrate to new debian key handling scheme, still downloading unverified
    3. add new optional check on gpg key fingerprint

LeSpocky commented:
Two things could be improved, I might wanna add those in v3:
- remove the old key with `apt-key del 9DC858229FC7DD38854AE2D88D81803C0EBFCD88`
- update documentation, README.md is still full of placeholders though, and could be improved independent from this PR

penguineer commented:
> * remove the old key with `apt-key del 9DC858229FC7DD38854AE2D88D81803C0EBFCD88`

Ansible is usually additive and additional keys don't really matter, so I would not bother with that.

> * update documentation, README.md is still full of placeholders though, and could be improved independent from this PR

Filling the gaps in README should be a different PR, but I would like the `docker_apt_key_fpr` variable to be documented somewhere (for now e.g. as a comment in  the defaults).

LeSpocky commented:
Just pushed v3 with these changes:

- fixed optional fpr check and added a documentation comment to *defaults/main.yml*
- added another commit for old apt-key removal: this one should actually remove the annoying warning about deprecated storage. reworded patch 2 commit message accordingly.

Will leave README update for later.

LeSpocky commented:
Tested in a different environment:

```
 _________________________________________________________
< TASK [netz39.host_docker : Add Docker's APT repository] >
 ---------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

task path: /home/alex/.ansible/roles/netz39.host_docker/tasks/main.yml:50
redirecting (type: lookup) ansible.builtin.passwordstore to community.general.passwordstore
redirecting (type: lookup) ansible.builtin.passwordstore to community.general.passwordstore
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ansible.errors.AnsibleUndefinedVariable: 'dict object' has no attribute 'deb'. 'dict object' has no attribute 'deb'
fatal: [miraculix]: FAILED! => {"changed": false, "msg": "AnsibleUndefinedVariable: 'dict object' has no attribute 'deb'. 'dict object' has no attribute 'deb'"}
```

Seems like `arch={{ ansible_local['deb']['arch'] }}` uses some things from the other environment.  Too bad I have no access on that until 2025-01-07 again. 😞
